osdc: refine Objecter's perf counter

houbin0504 opened the pull request:
Signed-off-by: houbin0504 houbinbj@inspur.com

Add perf counter for CEPH_OSD_OP_TMAPUP  in osdc,then we can count the numbers of this operation on each objecter send and each OSD received.It is easily to locate the reason of the problem that occured with those three operations.
Add 2 latency perf counters -- 1 for read ops and 1 for write ops. 

Fixes:http://tracker.ceph.com/issues/40564
Signed-off-by: houbin0504 <houbinbj@inspur.com>



<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test docs`
- `jenkins render docs`

</details>

batrick commented:
please use component name "osdc" in your commit title.

houbin0504 commented:
> please use component name "osdc" in your commit title.

OK,I already changed，thank you.

batrick commented:
See Sage's comment: https://tracker.ceph.com/issues/40564#note-2

dillaman commented on src/osdc/Objecter.cc:
```diff
@@ -2339,6 +2342,10 @@ void Objecter::_send_op_account(Op *op)
     case CEPH_OSD_OP_CMPXATTR: code = l_osdc_osdop_cmpxattr; break;
     case CEPH_OSD_OP_RMXATTR: code = l_osdc_osdop_rmxattr; break;
     case CEPH_OSD_OP_RESETXATTRS: code = l_osdc_osdop_resetxattrs; break;
+    case CEPG_OSD_OP_TMAPUP: code = l_osdc_osdop_tmap_up; break;
```
Typos

dillaman replied:
... this is still an issue causing a compile failure:

> /home/jenkins-build/build/workspace/ceph-pull-requests/src/osdc/Objecter.cc: In member function 'void Objecter::_send_op_account(Objecter::Op*)':
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/osdc/Objecter.cc:2344:10: error: 'CEPG_OSD_OP_TMAPUP' was not declared in this scope
>      case CEPG_OSD_OP_TMAPUP: code = l_osdc_osdop_tmap_up; break;
>           ^~~~~~~~~~~~~~~~~~
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/osdc/Objecter.cc:2344:10: note: suggested alternative: 'CEPH_OSD_OP_TMAPUP'
>      case CEPG_OSD_OP_TMAPUP: code = l_osdc_osdop_tmap_up; break;
>           ^~~~~~~~~~~~~~~~~~
>           CEPH_OSD_OP_TMAPUP
> src/osdc/CMakeFiles/osdc.dir/build.make:88: recipe for target 'src/osdc/CMakeFiles/osdc.dir/Objecter.cc.o' failed
> make[3]: *** [src/osdc/CMakeFiles/osdc.dir/Objecter.cc.o] Error 1

houbin0504 replied:
Sorry,It was my negligence.The file has been modified.

dillaman commented on src/osdc/Objecter.cc:
```diff
@@ -119,6 +119,9 @@ enum {
   l_osdc_osdop_rmxattr,
   l_osdc_osdop_resetxattrs,
+  l_osdc_osdop_tmap_up,
+  l_osdc_osdop_tmap_put,
```
TMAP_PUT and TMAP_GET have been removed from use so they will never provide any stats. RBD still uses TMAP_UP, though, for v1 format images.

houbin0504 replied:
Thank you very much for your revieded.I had removed the perf counter abort TMAP_PUT and TMAP_GET.

dillaman commented:
@houbin0504 It's still failing to compile due to typos. Also, can you please squash all your commits into a single commit and repush when fixed?

houbin0504 commented:
> @houbin0504 It's still failing to compile due to typos. Also, can you please squash all your commits into a single commit and repush when fixed?

I already squashed those three commits into a single commit and fixed it.

dillaman commented:
@houbin0504 Nit: the "Fixbug" keyword in your commit message should be "Fixes: ....", otherwise lgtm

houbin0504 commented:
> @houbin0504 Nit: the "Fixbug" keyword in your commit message should be "Fixes: ....", otherwise lgtm

I have updated according to your suggestion. Please review,thanks.

dillaman commented:
@houbin0504 I actually missed the second part of the tracker ticket where it's asking for read/write op latency stats. Can you also add new perf counters for that using `pcb.add_time_avg(...)`?

houbin0504 commented:
> @houbin0504 I actually missed the second part of the tracker ticket where it's asking for read/write op latency stats. Can you also add new perf counters for that using `pcb.add_time_avg(...)`?

I have already added them, please review,thank you.

dillaman commented on src/osdc/Objecter.cc:
```diff
@@ -377,6 +385,11 @@ void Objecter::init()
     pcb.add_u64_counter(l_osdc_osdop_omap_del, "omap_del",
 			"OSD OMAP delete operations");
 
+    pcb.add_time_avg(l_osdc_lat_big_write, "lat_big_write", "latency of write ops >= 128KB");
```
The 128KiB split is very arbitrary -- I would just rely on a single perf counter for reads and one for writes.

houbin0504 replied:
What do you think of adding a configuration item here?

dillaman replied:
I don't think anyone would ever touch it so it would again be arbitrary. Even the OSDs themselves don't segment op latencies into small and large buckets.

houbin0504 replied:
How do you think it's more appropriate to implement it here？Can you give me some suggestions?

dillaman replied:
I would just have 2 latency perf counters -- 1 for read ops and 1 for write ops.

houbin0504 replied:
I have updated according to your suggestion. Please review,thanks.

dillaman commented on src/osdc/Objecter.cc:
```diff
@@ -3166,6 +3181,30 @@ void Objecter::_finish_op(Op *op, int r)
+  int stat_type(0);
+  for (const auto& p : op->ops) {
+    switch (p.op.op) {
+      case CEPH_OSD_OP_READ:
```
Nit: align `case` with `switch` indentation above

houbin0504 replied:
I have modified it.Please review,thanks.

dillaman commented on src/osdc/Objecter.cc:
```diff
@@ -3166,6 +3181,30 @@ void Objecter::_finish_op(Op *op, int r)
+        stat_type = (p.op.extent.length >= (128 << 10)) ? l_osdc_lat_big_read : l_osdc_lat_small_read;
+      break;
+      case CEPH_OSD_OP_WRITE:
+      case CEPH_OSD_OP_WRITEFULL:
```
`_OP_APPEND` and `_OP_WRITESAME`?

houbin0504 replied:
I have added them.

dillaman commented on src/osdc/Objecter.cc:
```diff
@@ -3166,6 +3181,30 @@ void Objecter::_finish_op(Op *op, int r)
 
+  ceph::coarse_mono_time now = ceph::coarse_mono_time::now();
+  auto latency = now - op->init_stamp;
+  int stat_type(0);
```
Nit: `stat_type = 0;`.

houbin0504 replied:
I have modified it.
